lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,199 @@
+# Looks up data defined using Hiera and Data Providers
+# The function is callable with one to four arguments and optionally with a code block to provide default value.
+#
+# The lookup function can be called in one of these ways:
+#
+#     lookup(name)
+#     lookup(name, value_type)
+#     lookup(name, value_type, default_value)
+#     lookup(name, value_type, default_value, merge)
+#     lookup(options_hash)
+#     lookup(name, options_hash)
+#
+# The function may optionally be called with a code block / lambda with the following signature:
+#
+#     lookup(...) |$name| { ... }
+#
+# The block, if present, is mutually exclusive to the `default_value` and will be called with the `name` used in
+# in the lookup when no value is found. The value produced by the block then becomes the value produced by the
+# lookup.
+#
+# When a block is used, it is the users responsibility to call `error` if an undef value is not acceptable.
```

Comment:
It's still true since this PR doesn't have any logic to differ between not found and bound undef/nil. I expect that to be a separate issue.